smol/moca/composition/constraints.py

```diff
@@ -0,0 +1,298 @@
+            f" {num_numbers} > 1"
+            f" literals, not allowed!"
+        )
+
+    return number_literal, num_numbers
+
+
+def handle_side_string(side: str):
+    """Handle the strings separated by relation operators.
+
+    Args:
+        side(str):
+           A side of equation in string.
+    Returns:
+        list of tuple of (Number, Species, int or None), Number:
+            A list of tuples packing coefficients, species, and sub-lattice indices (if any)
+            on this side of equation, and the intercept term.
+    """
+    # split around spaces. each literal can be either "+", "-", "\d" or species string.
+    literals = side.split()
+    processed_literals = []
+
+    numerical_reg = re.compile(r"^[+-]?\d+\.?\d*$")
+    sublattice_reg = re.compile(r"^([A-Za-z]+.*)\((\d+)\)$")
+    for i, lit in enumerate(literals):
+        if lit == "+" or lit == "-":
+            processed_literals.append(lit)
+        elif numerical_reg.match(lit):
+            # might be a float.
+            if "." in lit:
+                num = float(lit)
+            # Integer
+            else:
+                num = int(lit)
+            # process to int if possible.
+            if np.isclose(num, round(num)):
+                num = round(num)
+            processed_literals.append(num)
+        else:
+            sublattice_match = sublattice_reg.match(lit)
+            # Specified a sub-lattice index.
+            if sublattice_match is not None:
+                species = get_species(sublattice_match.groups()[0])
+                sublattice_id = int(sublattice_match.groups()[1])
+            # No sub-lattice index.
+            else:
+                species = get_species(lit)
+                sublattice_id = None
+            processed_literals.append((species, sublattice_id))
+
+    # Pack literals and intercept term.
+    intercept = 0
+    if isinstance(processed_literals[-1], Number):
+        sym, n_sym = _extract_symbols_before_number(
+            processed_literals, len(processed_literals) - 1
+        )
+        intercept = sym * processed_literals[-1]
+        processed_literals = processed_literals[: -(1 + n_sym)]
+    elif processed_literals[-1] == "+":
+        intercept = 1
+        processed_literals = processed_literals[:-1]
+    elif processed_literals[-1] == "-":
+        intercept = -1
+        processed_literals = processed_literals[:-1]
+
+    packed_literals = []
+    # A species literal should never be preceded by more than 1 number literals.
+    for i, lit in enumerate(processed_literals):
+        if isinstance(lit, tuple):
+            # Extract the number literal before a species literal.
+            num, n_nums = _extract_number_before_species(processed_literals, i)
+            sym, _ = _extract_symbols_before_number(processed_literals, i - n_nums)
+
+            packed_literals.append((sym * num, *lit))
+
+    return packed_literals, intercept
```

Comment:
I still think this way of handling constraints is overly-verbose, I think we should implement a simpler programmatic input that needs less parsing, but I'll leave it to you to decide.

Reply:
This would be a trade-off between user-friendliness and programmer-friendliness. In older versions I used to allow dictionary inputs that requires zero parsing, but these formats would be very hard to interpret.  String inputs are straightforward to write and to interpret, one only needs to write expressions of species amounts, names and the corresponding sub-lattice index(in brackets); it's just the parsing that would require some effort. I prefer to make it easier for the users.

Reply:
As you can see, the vector input format is still kept so that the user can choose from one of the input methods.

Reply:
Ok, we can keep your current implementation. Do we have an example notebook that illustrates this? IF not it would be useful to have it shown there, otherwise it is buried in the documentation of the `CompositionSpace` class, which users rarely interact with

Reply:
Added setting-composition-constraints.ipynb.